src/components/TrackListItem.js

```diff
@@ -23,77 +23,81 @@ export const TrackListItem = ({
     availableColors,
     onChange, // expects a new trackProps object
     onDelete,
+    trackID,
   }) => {
-    const [myTrackProps, dispatch] = useReducer(reducer, trackProps);
-    const _onChange = useRef(onChange);
+    // propChanges only store new trackType, trackFile, and trackColorSettings changes
+    // reset after onChange is called
+    const [propChanges, setPropChanges] = useState({});
 
-    // https://overreacted.io/a-complete-guide-to-useeffect/
-    function reducer(state, action) {
-      let newState = {...state};
-      switch (action.type){
-        case "setType":
-          newState["trackFile"] = undefined;
-          newState["trackType"] = action.payload;
-          break;
-        case "setFile":
-          newState["trackFile"] = action.payload;
-          break;
-        case "setColorSettings":
-          newState["trackColorSettings"] = action.payload;
-          break;
-        default:
-          throw new Error();
-      }
-      return newState;
-    }
   
     const trackTypeOnChange = async(newTrackType) => {
       // wait until file change to call onChange
-      dispatch({type: "setType", payload: newTrackType});
+      let newPropChanges = {...propChanges};
+      newPropChanges["trackType"] = newTrackType;
+      newPropChanges["trackFile"] = undefined;
+      setPropChanges(newPropChanges);
     }
 
     const trackFileOnChange = async(newFile) => {
-      dispatch({type: "setFile", payload: newFile});
+      let newPropChanges = {...propChanges};
+      newPropChanges["trackFile"] = newFile;
+      setPropChanges(newPropChanges);
     }
 
     const trackSettingsOnChange = async(key, value) => {
       let newTrackColorSettings = {...trackProps["trackColorSettings"]};
       newTrackColorSettings[key] = value;
-      dispatch({type: "setColorSettings", payload: newTrackColorSettings});
+      let newPropChanges = {...propChanges};
+      newPropChanges["trackColorSettings"] = newTrackColorSettings;
+      setPropChanges(newPropChanges);
```

Comment:
Is it really better to set component state here and then have it re-render with the new state and then have useEffect see the state is safe to commit upwards and then have it call `onChange` and change the props and state again and then have it render again?

I feel like each event handler could call a `commitOrBufferPropChanges()` function with `newPropChanges`, and right inside the event handler we could decide if we are ready to call up to `onChange` or if we need to buffer the changes in `propChanges`, and that might be simpler and could save some renders.

But this looks like it ought to work the way it is.

Reply:
I was going down this path, but I ran into an issue where calling a setState hook and using its value afterward doesn't guarantee coherence. I'm not sure if I like the use of useEffect either, but it seemed like the consensus is that this works.

https://stackoverflow.com/questions/54069253/the-usestate-set-method-is-not-reflecting-a-change-immediately
https://stackoverflow.com/questions/54954091/how-to-use-callback-with-usestate-hook-in-react